daemon/lib/source/DobbyManager.cpp

```diff
@@ -1735,6 +1738,287 @@ bool DobbyManager::wakeupContainer(int32_t cd)
     return true;
 }
 
+// -----------------------------------------------------------------------------
+/**
+ *  @brief adds a mount to a running container
+ *
+ *  @param[in]  cd           The descriptor of the container to checkpoint.
+ *  @param[in]  source
+ *  @param[in]  destination
+ *  @param[in]  mountFlags
+ *
+ *  @return true if a container was successfully restored.
+ */
+bool DobbyManager::addMount(int32_t cd, const std::string &source, const std::string &destination, int32_t mountFlags)
+{
+     AI_LOG_FN_ENTRY();
+
+    std::lock_guard<std::mutex> locker(mLock);
+
+    // find the container
+    auto it = mContainers.cbegin();
+    for (; it != mContainers.cend(); ++it)
+    {
+        if (it->second && (it->second->descriptor == cd))
+            break;
+    }
+
+    if (it == mContainers.cend())
+    {
+        AI_LOG_WARN("failed to find container with descriptor %d", cd);
+        AI_LOG_FN_EXIT();
+        return false;
+    }
+
+    const ContainerId &id = it->first;
+    
+    const pid_t containerPid = it->second->containerPid;
+    if (containerPid == 0)
+    {
+        AI_LOG_WARN("failed to find container pid for container with descriptor %d", cd);
+        AI_LOG_FN_EXIT();
+        return false;
+    }
+    AI_LOG_INFO("containerPid: %d", containerPid);
+ 
+    // fork the process
+    pid_t pid = fork();
```

Comment:
Using fork is not ideal, presumably this was done because of the userns switch and setting the UID / GID for the `mkdir` command.  So I did an experiment and don't think it's necessary to switch into the userns which means I think you can implement this simply via the `IDobbyUtils::callInNamespace` API.

The trick is you need to switch into the uid / gid of the container, and you can do that with the `SYS_setresgid` and `SYS_setresuid` which set the effective UID / GID on a single thread.  So should be able to do the following:
```

int mountFd = open_tree(...

auto lambda = [mountFd, containerUid, containerGid]()
{
    // switch to uid / gid of the container
    if (syscall(SYS_setresgid, -1, containerUid, -1) != 0)
       AI_LOG_ERROR(....
    if (syscall(SYS_setresuid, -1, containerGid, -1) != 0)
        AI_LOG_ERROR(....

    mkdir(mountPoint ...);

    // switch back to root for the mount
    if (syscall(SYS_setresgid, -1, 0, -1) != 0)
        AI_LOG_ERROR(....
    if (syscall(SYS_setresuid, -1, 0, -1) != 0)
        AI_LOG_ERROR(....

    move_mount(fdMount, ...)
};

mUtilities->callInNamespace(pid, CLONE_NEWNS, lambda);

close(fdMount);
```

Reply:
@bgray-sky yes, I had to use fork() because of userns. I'll try your suggestion  which looks like a much cleaner way, thanks!

Reply:
I've removed the fork() and replaced it with setresgid/uid and  callinnamespace() calls. It works!